### PR TITLE
Add option to disable padding in rgblink

### LINKED ISSUE #307: Don't pad binaries with rgblink
I'm trying to use rgblink to link a small binary containing some data that was pre-processed using rgbds macros. This file is then passed through another program before finally being INCBINned in the final ROM.

The only problem to this method is that rgblink insists on padding the binary to 16K. It'd be nice to have an option to disable this padding, or have rgbfix handle padding to 16K boundaries instead.

### PR DESCRIPTION
**Pull request by JL2210**
Fixes #307

RGBFIX can handle padding, so there's no reason why
we can't add an option to disable padding in rgblink.

I don't think this is worth a short argument, though,
especially considering that all the good letters are used up.
Take up a letter for a short argument but don't put it in the `optstring`,
and leave a comment to this effect.

Signed-off-by: JL2210 <larrowe.semaj11@gmail.com>

### CODE REVIEW

**Comment by ISSOtm** [comment]
This is actually buggy, because it does not pad *between* banks, which RGBFIX thus cannot fix. This is why I didn't do this in response to #307.
The main problem is what one expects from an "unpadded" ROM? If it's 32k, I can understand, but less so when it's banked.

**Comment by JL2210** [comment]
Hmm... I see that. Maybe we could break the padding out of the function and not do it on the last iteration?

**Comment by ISSOtm** [comment]
Possible, but I'm not sure that's the expected behavior for multi-bank... not ROMs?

**Comment by JL2210** [comment]
Based on the requirements of this it might fit: https://github.com/rednex/rgbds/issues/307#issuecomment-431876032

Just need to add some checks for not using tiny mode.

**Comment by JL2210** [comment]
@ISSOtm What do you think?

**Comment by ISSOtm** [comment]
Hm, maybe. I'd like review from @aaaaaa123456789, though, since he's the one making the request.

**Comment by aaaaaa123456789** [comment]
I did? What does this feature relate to?

File: src/link/output.c [hunk 1]
```diff
@@ -175,9 +175,16 @@ static void writeBank(struct SortedSection *bankSections, uint16_t baseOffset,
 		bankSections = bankSections->next;
 	}
 
-	while (offset < size) {
-		putc(overlayFile ? getc(overlayFile) : padValue, outputFile);
-		offset++;
+	if (!disablePadding) {
```
Comment (aaaaaa123456789) [inline src/link/output.c:178]: I wouldn't particularly write an `if` that has both a negated condition and an `else`, but maybe that's just me.

File: src/link/main.c [hunk 2]
```diff
@@ -108,6 +109,8 @@ static struct option const longopts[] = {
 	{ "version",      no_argument,       NULL, 'V' },
 	{ "verbose",      no_argument,       NULL, 'v' },
 	{ "wramx",        no_argument,       NULL, 'w' },
+	/* x is reserved for nopad but is not a short argument */
```
Comment (aaaaaa123456789) [inline src/link/main.c:112]: If you're reserving `x` for no padding, why not just let it be a short argument?

File: src/link/main.c [hunk 3]
```diff
@@ -199,6 +203,9 @@ int main(int argc, char *argv[])
 		case 'w':
 			isWRA0Mode = true;
 			break;
+		case 'x':
```
Comment (aaaaaa123456789) [inline src/link/main.c:206]: I'd probably make no padding imply tiny mode, instead of enforcing it with a check.

Comment (aaaaaa123456789) [reply src/link/main.c:206]: Alternate implementation (would be more useful, but more work): no padding + no tiny mode works like tiny mode, but with an unbounded ROM0 size.

After all, $8000 is an irrelevant mark when the output is not a ROM.

File: src/link/main.c (hunk 2 above)
Comment (JL2210) [reply src/link/main.c:112]: I figured that `x` wouldn't be a very descriptive letter. All the others at least have some relation to their respective option but this one has none.

File: src/link/main.c (hunk 3 above)
Comment (JL2210) [reply src/link/main.c:206]: Considering that this was to be used as a temporary file to INCBIN into a ROM, I'd say that tiny mode would be very reasonable. I'll go ahead and make it imply tiny mode and remove the check.

File: src/link/main.c (hunk 2 above)
Comment (aaaaaa123456789) [reply src/link/main.c:112]: There's still no benefit in forbidding the short form if you're going to reserve the letter anyway.

Comment (NieDzejkob) [reply src/link/main.c:112]: If I'm reading the code correctly, the short option isn't even forbidden, but merely undocumented.

Note that you can use unprintable characters for this.

Comment (JL2210) [reply src/link/main.c:112]: @NieDzejkob It's not in the optstring, so it won't be recognized. Anyway, I have a commit ready to fix this.

File: src/link/output.c (hunk 1 above)
Comment (JL2210) [reply src/link/output.c:178]: Removed the `else` as per other comment.

File: src/link/main.c (hunk 2 above)
Comment (JL2210) [reply src/link/main.c]: done, in case you haven't noticed.

File: src/link/main.c (hunk 3 above)
Comment (JL2210) [reply src/link/main.c:205]: Implies tiny mode now, if you haven't seen it.

Comment (aaaaaa123456789) [reply src/link/main.c:205]: I did; I can't close the conversation or anything, though.

File: src/link/main.c (hunk 2 above)
Comment (aaaaaa123456789) [reply src/link/main.c]: It should've been automatically folded by GitHub, since the code is now marked as outdated.

**Comment by ISSOtm** [comment]
The code LGTM, but could you also update the documentation? (`src/link/rgblink.1`, specifically)

**Comment by JL2210** [comment]
I have no experience with writing manpages whatsoever, so probably not.

**Comment by ISSOtm** [comment]
Ok, then I'll do that myself.

**Comment by ISSOtm** [comment]
Bless "Allow edits from maintainers"! Merging.

**Pull request merged**